internal/server/audit/audit.go

```diff
@@ -0,0 +1,244 @@
+package audit
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/trace"
+	"go.uber.org/zap"
+)
+
+const eventVersion = "v0.1"
+
+// Type represents what resource is being acted on.
+type Type string
+
+// Action represents the action being taken on the resource.
+type Action string
+
+const (
+	Constraint   Type = "constraint"
+	Distribution Type = "distribution"
+	Flag         Type = "flag"
+	Namespace    Type = "namespace"
+	Rule         Type = "rule"
+	Segment      Type = "segment"
+	Variant      Type = "variant"
+
+	Create Action = "created"
+	Delete Action = "deleted"
+	Update Action = "updated"
+)
+
+// Event holds information that represents an audit internally.
+type Event struct {
+	Version  string      `json:"version"`
+	Metadata Metadata    `json:"metadata"`
+	Payload  interface{} `json:"payload"`
+}
+
+// DecodeToAttributes provides a helper method for an Event that will return
+// a value compatible to a SpanEvent.
+func (e Event) DecodeToAttributes() []attribute.KeyValue {
+	akv := make([]attribute.KeyValue, 0)
+
+	if e.Version != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.version",
+			Value: attribute.StringValue(e.Version),
+		})
+	}
+
+	if e.Metadata.Action != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.metadata.action",
+			Value: attribute.StringValue(string(e.Metadata.Action)),
+		})
+	}
+
+	if e.Metadata.Type != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.metadata.type",
+			Value: attribute.StringValue(string(e.Metadata.Type)),
+		})
+	}
+
+	if e.Metadata.IP != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.metadata.ip",
+			Value: attribute.StringValue(e.Metadata.IP),
+		})
+	}
+
+	if e.Metadata.Author != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.event.metadata.author",
+			Value: attribute.StringValue(e.Metadata.Author),
+		})
+	}
+
+	if e.Payload != nil {
+		b, err := json.Marshal(e.Payload)
+		if err == nil {
+			akv = append(akv, attribute.KeyValue{
+				Key:   "flipt.event.payload",
+				Value: attribute.StringValue(string(b)),
+			})
+		}
+	}
+
+	return akv
+}
+
+func deserializeHelper(s string) string {
+	ss := strings.Split(s, ".")
+	return ss[len(ss)-1]
+}
+
+func (e *Event) Valid() bool {
+	return e.Version != "" && e.Metadata.Action != "" && e.Metadata.Type != "" && e.Payload != nil
+}
+
+var errEventNotValid = errors.New("audit event not valid")
+
+// decodeToEvent provides helper logic for turning to value of SpanEvents to
+// an Event.
+func decodeToEvent(kvs []attribute.KeyValue) (*Event, error) {
+	e := new(Event)
+	for _, kv := range kvs {
+		terminalKey := deserializeHelper(string(kv.Key))
+
+		switch terminalKey {
```

Comment:
Thought: Maybe we should actually match on the entire key instead.
Just wondering if there is a chance these collide with other event keys?

This kind of removes the namespacing feature of `flipt.area.` if we split on `.` and only compare the last entry.

Reply:
@GeorgeMac Really good thinking. I agree, and make that change to match on the whole key.